### PR TITLE
test, fix deepcopy of tensor with grad

### LINKED ISSUE #3307: `copy.deepcopy` does not copy gradient buffers of `torch.autograd.Variable` instance
I ran into unexpected behaviour with `copy.deepcopy` applied to a `Variable`. The gradient buffer of the `Variable` is not copied.
```
a = torch.autograd.Variable(torch.ones(1))
a.grad = torch.autograd.Variable(torch.ones(1))
b = copy.deepcopy(a)
print(b.grad)
```
I think it would be a good idea to copy the gradient buffer during a deep copy. My use case is recording the gradient of a model's parameter space for optimization research. This would also be useful for debugging/development of complex models that involve atypical gradient operations.

This is handled here: https://github.com/pytorch/pytorch/blob/5760b036fb338eacd641418321f23aee51b1aee9/torch/autograd/variable.py#L89-L97

A solution would be to also copy the `grad` attribute of the current `Variable`, which would involve a recursion of the deep copy since the grad attribute is also a `Variable`.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @gqchen @pearu @nikitaved @soulitzer @SsnL

### PR DESCRIPTION
**Pull request by mattip**
Fixes #3307

Previously, `self.grad` was not ~cloned~ deepcopied to the returned tensor in `deepcopy`. Added a test and an implementation.

### CODE REVIEW

File: torch/tensor.py [hunk 1]
```diff
@@ -77,6 +77,8 @@ def __deepcopy__(self, memo):
                     new_tensor = self.new()
                     new_tensor.set_(new_storage, self.storage_offset(), self.size(), self.stride())
                     new_tensor.requires_grad = self.requires_grad
+            if self.grad is not None:
+                new_tensor.grad = self.grad.clone()
```
Comment (albanD) [inline torch/tensor.py:81]: This should be deepcopied instead of cloned no? Otherwise you might loose things like sharing?

File: test/test_torch.py [hunk 2]
```diff
@@ -2881,6 +2881,17 @@ def test_deepcopy_scalar(self, device, dtype):
         self.assertEqual(a.size(), deepcopy(a).size())
         self.assertEqual(a, deepcopy(a))
 
+    @dtypes(torch.float32)
+    @onlyCPU
```
Comment (albanD) [inline test/test_torch.py:2885]: nit: why CPU only?

File: test/test_torch.py [hunk 3]
```diff
@@ -2881,6 +2881,17 @@ def test_deepcopy_scalar(self, device, dtype):
         self.assertEqual(a.size(), deepcopy(a).size())
         self.assertEqual(a, deepcopy(a))
 
+    @dtypes(torch.float32)
```
Comment (albanD) [inline test/test_torch.py:2884]: Note that if you just want a single test, you can just more it to `TestTorch` above and it will be called with no special argument and you can specify the type there directly.

**Review by albanD — COMMENTED** [review]
Thanks for the PR!

I added comments inline below.

File: test/test_torch.py (hunk 3 above)
Comment (mattip) [reply test/test_torch.py:2884]: moved to TestTorch

File: test/test_torch.py (hunk 2 above)
Comment (mattip) [reply test/test_torch.py]: moved to TestTorch

File: torch/tensor.py (hunk 1 above)
Comment (mattip) [reply torch/tensor.py]: fixed

**Comment by mattip** [comment]
@albanD updated

**Review by albanD — COMMENTED** [review]
Thanks for the update. Just one small addition to the test and this is good to land!

File: test/test_torch.py [hunk 4]
```diff
@@ -7055,6 +7044,16 @@ class TestTensorDeviceOps(TestCase):
...
     exact_dtype = True
 
+    def test_deepcopy_gradient(self):
+        from copy import deepcopy
+        a = torch.zeros(10)
+        a.grad = torch.ones(10)
+        self.assertEqual(a.grad, deepcopy(a).grad)
+        s = torch.zeros(10).to_sparse()
+        s.grad = torch.ones(10).to_sparse()
+        self.assertEqual(s.grad, deepcopy(s).grad)
```
[the remark this answers is not in the document]
Comment (mattip) [reply test/test_torch.py:7054]: done

**Comment by mattip** [comment]
Not sure what is up with [CI](https://app.circleci.com/pipelines/github/pytorch/pytorch/262553/workflows/f4aa06fd-307a-4c77-bb67-663f66f4664c), but I don't think this PR is causing the failures

**Comment by albanD** [comment]
Thanks for the update!
I am not sure if the CI error are related. But master is all green, so I think this failure is introduced by this PR.

I think the issue is that this test [here](https://github.com/pytorch/pytorch/blob/4ac489091ab2e9ee34ca6da4ee7327b76e646923/torch/testing/_internal/common_nn.py#L5098) now copies the grad of the input which is non-zero. So we should zero it out before calling the backward just like we do for `input`.

**Comment by mattip** [comment]
> So we should zero it out before calling the backward just like we do for input.

Like in 9ec84d0 ?

File: torch/tensor.py [hunk 5]
```diff
@@ -77,6 +78,8 @@ def __deepcopy__(self, memo):
                     new_tensor = self.new()
                     new_tensor.set_(new_storage, self.storage_offset(), self.size(), self.stride())
                     new_tensor.requires_grad = self.requires_grad
+            if self.grad is not None:
+                new_tensor.grad = deepcopy(self.grad)
```
Comment (albanD) [inline torch/tensor.py:82]: I think that for the sharing to work, you need to call this the same way the storage copy is called above: `new_tensor.grad = self.grad.__deepcopy__(memo)`.

If you do that change, don't forget to remove the import for `deepcopy` to make sure the lint is happy.

**Review by albanD — COMMENTED** [review]
> Like in 9ec84d0 ?

Yes that seems to have fixed the test.

Some test now fail because there are merge conflict with master, you can rebase on top of master and re-push to solve these.
Also the check for sharing seems to fail. See my comment below on why I think it is expected.

File: torch/tensor.py (hunk 5 above)
Comment (mattip) [reply torch/tensor.py]: done

**Review by albanD — APPROVED** [review]
All green! Thanks!
